### ISSUE TITLE
Interoperability with Ivy and Maven dependency resolvers

### ISSUE THREAD

**Issue by ckulenkampff** [ceylon/ceylon-herd#262]
This enhancement would allow to create a flat classpath of Ceylon CARs for Java projects using Maven, Ivy or Gradle. This is possible by offering appropriate repository "facades" through the Herd repository server.

**Maven repository structure**
Maven expects the following layout (see [Maven Repository Layout - Final]) for primary artifacts:
`/$groupId[0]/../${groupId[n]/$artifactId/$version/$artifactId-$version.$extension`
and for secondary artifacts:
`/$groupId[0]/../$groupId[n]/$artifactId/$version/$artifactId-$version-$classifier.$extension`

**Ivy repository structure**
Ivy is more flexible and allows to specify custom patterns for artifact resolution (see [Ivy Documentation - Main Concepts]). The default patten that is used by Gradle is the following (see [Gradle DSL Reference - IvyArtifactRepository]):
Artifacts: `$baseUri/[organisation]/[module]/[revision]/[type]s/[artifact](.[ext])`
Ivy module descriptors: `$baseUri/[organisation]/[module]/[revision]/[type]s/[artifact](.[ext])`

**Meta information**
To resolve transitive dependencies both repository types require meta information. Maven uses `pom.xml`s. Ivy uses `ivy.xml`s, but can also process `pom.xml`s. Those files must be accessible via HTTP requests.

**Meta information augmentation**
When the repository server responds to a "foreign" meta data request for a Ceylon module, it should automatically add all implicit dependencies of the Ceylon language to the response. For interoperability these Ceylon language modules should be published to the Herd repository so that Java projects that depend on a Ceylon library do not have to provide them by themselves.

**Artifact aliases**
For interoperability it would be very useful when CAR files are also available under the same name but with JAR file extension when accessed through a facade.

Many IDEs automatically link source and javadoc JARs to the downloaded artifacts by searching in the module cache for files like `$artifactId-$version-$classifier-sources.$extension` (IDE dependent see [NetBeans DependencyNode]). Ceylon source artifacts should be made available in a way that this resolution works out of the box. This means that the artifacts are made available under another name than they are normally accessible in Herd.

[Maven Repository Layout - Final]: https://cwiki.apache.org/confluence/display/MAVENOLD/Repository+Layout+-+Final
[Ivy Documentation - Main Concepts]: http://ant.apache.org/ivy/history/latest-milestone/concept.html#patterns
[Gradle DSL Reference - IvyArtifactRepository]: https://docs.gradle.org/current/dsl/org.gradle.api.artifacts.repositories.IvyArtifactRepository.html#N217AD 
[NetBeans DependencyNode]: https://hg.netbeans.org/core-main/file/default/maven/src/org/netbeans/modules/maven/nodes/DependencyNode.java#l483

**Comment by FroMage** [comment]
I had the same thought the other day, I think that'd be really handy indeed. Would you be willing to test this for me?

**Comment by FroMage** [comment]
So Maven repo docs fail to mention these:

- http://central.maven.org/maven2/last_updated.txt (simple enough timestamp)
- view-source:http://central.maven.org/maven2/archetype-catalog.xml (not applicable: archetype list)
- http://central.maven.org/maven2/org/hibernate/maven-metadata.xml (not applicable: maven plugin list)
- http://central.maven.org/maven2/org/hibernate/hibernate-core/maven-metadata.xml (list of versions per groupId:artifactId)

**Comment by FroMage** [comment]
The biggest problem would be that Ceylon modules don't have a group/artifact split. We can emulate one, by splitting on the last `.`, but it may be weird.

**Comment by bjansen** [comment]
That's already what happens in the `pom.xml` stored in every car file:

```xml
<?xml version="1.0" ?>
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
 <modelVersion>4.0.0</modelVersion>
 <groupId>ceylon.interop</groupId>
 <artifactId>java</artifactId>
 <version>1.2.2</version>
 <name>ceylon.interop.java</name>
 <dependencies>
  <dependency>
    <groupId>ceylon</groupId>
    <artifactId>collection</artifactId>
    <version>1.2.2</version>
  </dependency>
 </dependencies>
</project>
```

**Comment by bjansen** [comment]
Alternatively, we could add an annotation in `module.ceylon`, something like:

```ceylon
mvn("ceylon", "interop.java")
module ceylon.interop.java 1.0.0 {
    ...
}

**Comment by gavinking** [comment]
> We can emulate one, by splitting on the last `.`, but it may be weird.

Why can't the group and artifact be identical: the module name?

**Comment by luolong** [comment]
That's a slippery slope right there -- adding Maven specific annotations to language module...

Maybe have a `group` annotation instead. This could be an informative tag for Herd, that could be used as a grouping of similar modules together.

And additionally, this would be used for mvn pom generation.

Lacking `group` annotation, group and artifact id could very well be module's full name

**Comment by ckulenkampff** [comment]
Usually the groupid refers to the project. See http://central.sonatype.org/pages/choosing-your-coordinates.html 
> The groupId identifies your project uniquely across all projects and you control this section of the overall name-space.

I think a better default would be:
groupId = Ceylon module name
artifactId = last component of the Ceylon module name

> Why can't the group and artifact be identical: the module name?

This might be the best default, because then the artifact would have already the right name.

**Comment by ckulenkampff** [comment]
> Would you be willing to test this for me?

Yes. I will try to create some kind of integration test for this.

**Comment by FroMage** [comment]
@vietj: what do you think we should do about group/artifact?

**Comment by FroMage** [comment]
Note that if we change the group/artifact mapping in Herd, we will also want to change it in the generated `pom.xml` in the `.car` files…

**Comment by bjansen** [comment]
I think Herd should simply extract the `pom.xml` from the car file, and build the correct hierarchy of folders + generate checksums, that should be enough to expose a Maven repo, right?

**Comment by FroMage** [comment]
That is indeed another option. Except that:

- some files (jars) may not have any `pom.xml`
- Herd has more info than the `pom.xml` files, although cosmetic, such as authors, urls, licence, description, etc…

**Comment by davidfestal** [comment]
So it should be a mix of both, as we do when generating OSGI metadata in manifest: reuse information specified in the internal POM, and add any additional information that can be provided by Herd.

**Comment by bjansen** [comment]
> reuse information specified in the internal POM, and add any additional information that can be provided by Herd

Looks like a good compromise.

> authors, urls, licence, description, etc…

Technically, poms can also contain such data, so we could fill it from `module.ceylon` if present (is Herd already doing this?).

**Comment by FroMage** [comment]
I can, that's why I said that it's a bit richer if I generate the `pom.xml` rather than use the one inside the `.car`.

**Comment by FroMage** [comment]
https://modules.ceylon-lang.org/maven/1/ceylon/language/1.2.1/language-1.2.1.pom
https://modules.ceylon-lang.org/maven/1/ceylon/language/maven-metadata.xml

Can you guys try it out?

**Comment by FroMage** [comment]
Supports `.jar`, `.jar.sha1`, `-sources.jar`, `.-sourcesjar.sha1`, `.pom`, `.pom.sha1` and `maven-metadata.xml`.

Does not support javadoc yet but I'm pretty sure the Java tools would not be able to make sense of ceylondoc anyway.

**Comment by ckulenkampff** [comment]
Wow so fast! I did a small test with Gradle. It works like a charm! I think the `maven-metdata.xml`s are primarily for interaction with other repository servers. So I will try to add the Ceylon Herd server to a Sonatype Nexus server.

I will have more time tomorrow then I will give you more detailed feedback. Do you think it's useful to have some kind of Gradle integration tests that can be run against the Herd server?

**Comment by FroMage** [comment]
Yeah, I think so too, which is why I asked @vietj about his opinion.

**Comment by vietj** [comment]
having an annotation to specify a group id : can raise a problem if you want to deduce de GAV from the Ceylon module name because you don't have the information

**Comment by vietj** [comment]
having a scheme with multiple group ids can make a problem later if you want to put the same deps on maven central because usually you owns a single group id

**Comment by vietj** [comment]
at the same time, maven does not manage cyclic dependencies, so I don't see how you would publish ceylon cyclic dependencies in a maven repo.

**Comment by FroMage** [comment]
ATM @vietj has parts of the distrib published at http://mvnrepository.com/artifact/org.ceylon-lang under the `org.ceylon-lang` group.

Using `ceylon.language:ceylon.language` as coordinates would make publishing to Maven Central harder has every groupId has to be registered (same as on Herd BTW, and we don't see that as a big problem although we do have an issue open to claim _domains_ or wildcards).

**Comment by FroMage** [comment]
Another option is to use a common groupId for every Ceylon module: `ceylon:ceylon.language`.

**Comment by renatoathaydes** [comment]
why not use the same coordinates that already are in [maven central](http://search.maven.org/#search%7Cga%7C1%7Cg%3A%22org.ceylon-lang%22)?

**Comment by ckulenkampff** [comment]
But what is about other projects? Let's assume somebody wants to publish to Maven Central and Herd. In this situation it would be very important to be able to control artifactid and groupid. Otherwise the dependencies might be downloaded from the Maven repository and the Herd repository and both would get in the classpath.

I begin to think that the best idea is to have an annotation that ensures compatibility. Ivy, Maven and many other tools use at least group id and artifact id for resolution. For Ceylon modules an annotation for group id might be enough, but when people want to upload JARs to Herd _and_ Maven they may want to use their existing groupid and artifact-id-scheme.

So maybe an annotation that can override group id and artifact id for repositories that support both coordinates would be the most interoperable way. The developer has to decide if she wants to use "Herd-mode" (Module-Version) or "Maven-mode" (Group-Module-Version) for module resolution.

**Comment by gavinking** [comment]
> why not use the same coordinates that already are in [maven central](http://search.maven.org/#search%7Cga%7C1%7Cg%3A%22org.ceylon-lang%22)?

I agree that `org.ceylon-lang` feels natural.

Here's another idea: how about `org.ceylon-lang.modules`? i.e. the address of modules in Herd?

**Comment by renatoathaydes** [comment]
If you don't change the group, you don't need to go through the sonatype process again to determine if you own the groupId or not.

**Comment by ckulenkampff** [comment]
> Here's another idea: how about org.ceylon-lang.modules? i.e. the address of modules in Herd?

Imagine a big project that uses a Ceylon module which itself depends on a Java-Jar module uploaded to Herd (possible?). The Ceylon module was developed only with Herd, so it is refered in the pom as `org.ceylon-lang.modules:com.company.project.module`. The big project uses the same library itself, but uses the normal Maven project coordinates. Now the same module is mentioned twice: `com.company.project:module` and `org.ceylon-lang.modules:com.company.project.module`.

With an repository interoperability annotation this problem would not occur: When Herd simulates a Maven repository it would always use the coordinates specified by the annotation (also for transitive dependencies).

The only question that would remain is, how the groupid is, when the module does not use the interoperability annotation.

I think in the future in the light of Jigsaw Maven and Ivy artifact resolvers will support multi-version resolution of dependencies. So I think finding a good solution for mapping artifacts between Herd and Maven/Ivy is crucial.

**Comment by renatoathaydes** [comment]
> I think in the future in the light of Jigsaw Maven and Ivy artifact resolvers will support multi-version resolution of dependencies

We already went through this a few times, but here we go again: no, Jigsaw does not, and according to the developers working on it, has no goal of, supporting multiple versions of a module at runtime. Jigsaw does not even include any version resolution strategy at all.

In my opinion, the solution is simple: add an optional groupId to Ceylon modules in general. If a module has one, it may be possible to publish it on Maven repos as well as Ceylon ones, otherwise it just can't be published in a Maven/Ivy repo.

To solve the problem of where Ceylon will look for when resolving a module, I would say, keep the current scheme mostly the same and:

* look first in the local Ceylon repos
* try Herd
* try other Ceylon-specific repos
* if the module has a groupId:
* * look at Maven repos the user may have configured using eather

Which repo a module comes from should not matter, but Ceylon should just prefer Ceylon-specific first. There's no need for the esoteric solutions described by @ckulenkampff

**Comment by ckulenkampff** [comment]
> > I think in the future in the light of Jigsaw Maven and Ivy artifact resolvers will support multi-version resolution of dependencies

> We already went through this a few times, but here we go again: no, Jigsaw does not, and according to the developers working on it, has no goal of, supporting multiple versions of a module at runtime. Jigsaw does not even include any version resolution strategy at all.

Sorry, didn't knew that. That made me wonder and I just looked up the specs. Seems to be vendor specific (see bottom of http://openjdk.java.net/projects/jigsaw/spec/reqs/). So I guess the JBoss Module System will become a valid Jigsaw Module system and it already supports multiple versions... But even if this is irrelevant, I still think a good mapping is very important.

> Back to ceylon modules...

My use case is to use Ceylon modules in Java projects.

> no need for the esoteric solutions

This would be great!

I just think that underestimating the possible problem space is dangerous. I think it's just a matter of time when a sophisticated hack hits Maven Central or a similar repository hard (see http://branchandbound.net/blog/security/2012/03/crossbuild-injection-how-safe-is-your-build/). 

A Maven interop specific group id should at least result in an administrative process where somebody has to verify the group id in addition to the Ceylon module name... And exactly this would be the purpose of an ownership key placed in the pom. It would bring bad light on Herd when the artifacts there collide with somebody else's group ids on Maven Central or jCenter.

**Comment by ckulenkampff** [comment]
I think the mapping is also important because of this situation:
The following situation is not that far fetched:
```
Ceylon Project --> Ceylon Module on Herd --> Artifact A on Maven Central --> Artifact B on Maven Central
                                         --> Artifact B on Herd
```

**Comment by renatoathaydes** [comment]
@ckulenkampff the above situations are not a problem at all... with any build system, in which repo the artifact can be found is completely irrelevant, if they have the same coordinates they are the same...

**Comment by luolong** [comment]
I guess we can't really get around the problem of specifying group id's for a module if and when we want to publish those modules as maven artifacts.

As creation of POM is already part of compiler, this decision has to be done on a (backend) compiler level. I suggest that since maven group id is essentially a java backend compiler concern, why not make it an official compiler configuration option.

Have the default behavior to be as it currently is - where module name is split at the last dot – everything on the left of the last dot will become _Maven Group id_ and the last segment will become artifact name.

Ex: Module `ceylon.interop.java/1.2.2` would become `ceylon.interop:java:1.2.2:car`.

Alternatively we could keep the full module name as maven artifact name and only use everything before the last dot as group id.

Like this: `ceylon.interop:ceylon.interop.java:1.2.2:car`

Additionally have a set of compiler flags for ceylon compile tool `--maven-group=ceylon-sdk` and `--maven-name=interop.java` (or something like this) for overriding the default.

These could also be read from a `.ceylon` configuration file of the project.

Then there would be no need for any backend specific `group` annotation in the language module.

**Comment by ckulenkampff** [comment]
This seems to me to be a good solution. The administrative issues will be there anyway. Will the CMR also look into the pom to identify artifacts that are already as Herd modules on the classpath?

**Comment by gavinking** [comment]
@luolong It seems that `ceylon.interop:java` would be more consistent with how stuff like Hibernate and Spring do it today.

But there's a problem with your suggestion: `ceylon.interop.java` gets the group id `ceylon.interop`; the rest of the SDK gets the group id `ceylon`. That doesn't look quite right.

**Comment by ckulenkampff** [comment]
You could specify maven name and maven group for the ceylon projects as well?
Something like this might be sufficient: `org.ceylon-lang:interop-java`

**Comment by quintesse** [comment]
I could accept the `--maven-group` part of @luolong 's suggestion. That `ceylon.interop.java` gets treated differently from the rest of the SDK just goes to show that we really need a way to override the default behaviour. So the build scripts in the SDK would just compile with `--maven-group=org.ceylon-lang`.

But I'm not so sure about `--maven-name` though, the compiler normally deals with multiple modules at the same time. So we would need give an error when you use that option with multiple modules. But in general it's not an option that really "adjusts" well to normal projects. (What if you have it in your ceylon config file and you add a new module? Unless we start extending the config to be able to set options based on the module name)

So in the end I'm back to thinking that to me a group/artifact name is an integral part of a module, so personally I feel more for an `artifact("org.ceylon-lang:interop-java")` annotation on `module` to be honest.

**Comment by ckulenkampff** [comment]
> But I'm not so sure about --maven-name though, the compiler normally deals with multiple modules at the same time.

damn

**Comment by gavinking** [comment]
So yeah, I was actually thinking along the lines of "well, this is actually rightly a _project_-level setting, _not_ a module-level setting". Unfortunately we don't really have a very well-defined notion of what a project is right now. We *do* have the notion, implicitly, in the CLI, and explicitly, in the IDEs, but we don't have a place to specify project-level metadata, except as compiler settings.

**Comment by bjansen** [comment]
> we don't have a place to specify project-level metadata

What about `.ceylon/config` and `.ceylon/ide-config`? They apply to all the modules under that directory.

**Comment by quintesse** [comment]
> this is actually rightly a project-level setting

Partly yes. But that would only work if you can still figure out a way to handle the name part. Sure we could go with some useful default but there will always be people we really really *really* want their artifact name to be different from the default.

**Comment by ckulenkampff** [comment]
As far as I understand one project can contain multiple modules, which get published separately. Gradle anticipates Jigsaw in a similar way (see https://docs.gradle.org/current/userguide/java_software.html#defining_api). Don't you want to specify group and artifact id per module? As @quintesse explained you may want to use org.ceylon-lang as project wide group id and use Ceylon module specific artifact ids.

**Comment by FroMage** [comment]
> with any build system, in which repo the artifact can be found is completely irrelevant, if they have the same coordinates they are the same...

Well… That's not entirely true. It depends. What we're trying to do is to make it so that the Maven coordinates could be the same on Herd and Maven Central. The Ceylon module name will remain the same.

BUT there are two sides to resolving. Maven/Gradle users that resolve things will use Maven coordinates. Ceylon module descriptors on the other hand will mostly use Ceylon module names, and those are not resolved using Maven resolvers. Only module names containing `:` will be resolved using the Maven resolver.

So for example, in Ceylon, importing `ceylon:collection` and `ceylon.collection` will get you two different modules. The first will be resolved using the Maven resolver and its embedded `pom.xml` and the second will be resolved using the Ceylon resolver and its embedded module descriptor.

It's not clear how we can avoid situations where people use Maven to resolve a Ceylon module using its Maven coordinates/repo, which will drag its dependencies using the Maven coordinates too, and then when calling the compiler/runner, Ceylon will use the Ceylon module descriptor and download Ceylon modules and work with those.

It makes at least for a double download.

Now, running using the `Main` API is a bit different in that Ceylon will not download any modules, but it will initialise the metamodel using metadata found in the classpath. Pretty sure that even for Ceylon modules fetched by Maven it will not use the `pom.xml` first, but the Ceylon module descriptor.

I'm not saying there's a problem there. I'm saying it's a little more complex than may seem, and not obvious at all works out.

**Comment by gavinking** [comment]
> Don't you want to specify group and artifact id per module?

I can't imagine why you would want to have different group ids for modules in the same project. That sounds wrong.

**Comment by gavinking** [comment]
And, to be clear, I _really_ don't want to see maven metadata polluting the module descriptors for the Ceylon SDK, OK? I need you guys to focus your minds on that requirement before you start trying to add `groupId` and `artifactId` annotations all over the SDK.

**Comment by quintesse** [comment]
> I can't imagine why you would want to have different group ids for modules in the same project. That sounds wrong.

Don't tell people they're wrong, it just gives them more incentive to go and actually want to do it ;)

But I can imagine very well that large projects will have modules divided up in groups. Now sure, you can then also put them in separate source folder and then run the compiler several times for each group with different arguments. But that all goes a bit against the nice thing we have now where often you don't even need a build script because typing `ceylon compile` in the root of a project will just correctly compile everything.

>  I really don't want to see maven metadata polluting the module descriptors for the Ceylon SDK

I don't see how we can prevent this information from leaking out. Sure we can choose our defaults in such a way that we don't need to add anything to our SDK, but other people will surely come up with examples where they *do* need more exact control over those names.

**Comment by FroMage** [comment]
> I really don't want to see maven metadata polluting the module descriptors

That's OK, we'll only add _Gradle_ metadata ;)

**Comment by FroMage** [comment]
Or _Ivy_ metadata, that's cool, right?

**Comment by gavinking** [comment]
> Don't tell people they're wrong

They're wrong. There. I told them.

Seriously. This path you want to go down is starting to scare me, and turn me off the whole idea. If you can come up with a solution that _doesn't_ involve doing nasty horrible polluting things and inventing some over-flexible system designed to accommodate every possible whim that somebody might someday have, then I'm much more open to the idea.

**Comment by gavinking** [comment]
Remember: group id and artifact id are two totally arbitrary strings with no interesting semantics. The group id could be `"foo"` for all Ceylon modules and everything would still work.

**Comment by quintesse** [comment]
> with no interesting semantics

That's not true, for Maven and Ivy etc this is what defines your *ownership*.
You can't just go assigning arbitrary strings.

**Comment by renatoathaydes** [comment]
@gavinking correct in the Ceylon world, but I can't publish to Maven central under any other groups than the one(s) I own... so I need to declare my groupID... I have to say I agree with Gavin, go for the simplest possible solution that may work.. which seems to be that, when in the Maven world, groupId is anything before the last `.` in the module name? so no magic in the Ceylon side.... if people want magic, I could add some magic to the Gradle plugin so they can publish poms to Maven repos with any groupdId:artifactId they want... and the Maven plugin could do the same.

**Comment by quintesse** [comment]
@renatoathaydes that might cause problems when what you published in Maven has a different name of that what you published on the Herd. Eg. for the SDK modules on the Herd we won't know that they're called `org.ceylon.lang:xxx` in Maven. So they become different modules when obtained via Maven or via the Herd. That doesn't seem like a good idea.

**Comment by renatoathaydes** [comment]
@quintesse  Perhaps you should consider using different rules for Ceylon's SDK modules and people's modules... but if people decide to publish things with different names in different repos, they will have problem with any build system in the world.

**Comment by quintesse** [comment]
>  but if people decide to publish things with different names in different repos

@renatoathaydes exactly the reason why putting this information in the module descriptor would be a good idea ;)

In this case it's not them using different names, it's us. If they put their modules in Maven using their officially registered group name and also publish to Herd where *we* will possible auto-generate a different group name according to our own rules then no matter what they do they're up shit creek.

**Comment by renatoathaydes** [comment]
Tricky... seems like the only way to avoid trouble is to add a `group` annotation to the module descriptor... and if the module has no such annotation, it won't be considered for look up in Maven repos?

**Comment by renatoathaydes** [comment]
instead of a `group` annotation, maybe the current solution to declare Maven dependencies with a String as in `"group:name" "1.0"` could be considered as well... for consistency sake.

**Comment by ckulenkampff** [comment]
Yeah, a decision will severely affect Java-Ceylon-interaction in the long run.

For what purpose is the pom.xml in the CAR generated by Ceylon in the first place?

Maybe the best way to deal with it, is to not generate the pom.xml automatically and to ditch the Maven interoperability in Herd that I proposed here.

All projects that want to be available in Maven would have to add an additional publishing process to their build that generates a pom file and uploads their CAR as JAR to Maven. To make this work you would have to publish Ceylon core stuff also to Maven.

**Comment by ckulenkampff** [comment]
Just for the record, the artifact id is as important as the group id. Both can not safely deduced from the Ceylon module name. They are simply two different things. The Ivy Maven interop works so well because they have organization and module name as coordinates. See http://ant.apache.org/ivy/history/latest-milestone/ivyfile.html

Should we add another issue that deals with Maven / Ceylon artifact coordinate interoperability in another project issue tracker?

**Comment by gavinking** [comment]
I think what everyone is forgetting here is that the person writing and compiling the module is not the person who wants to depend on the module using mvn. Therefore a system which depends on the module author (optionally) annotating the module at compile time is unlikely to work well in practice.

**Comment by bjansen** [comment]
Perhaps this has already been discussed in a previous comment, but we're talking about serving a Maven repo from Herd, so why not configure the group id/artifact id in Herd directly? This can be during the project claim, or during the project upload, with default values extracted from the module name.

pros:
* can be easily modified on Herd
* no need to add annotations to `module.ceylon`
* Herd can generate a new pom, based on all the project settings we fill during the claim/upload
* we don't need to recompile existing modules just to add metadata

cons:
* there's a lot of already existing projects on Herd, so we'll have to add maven coordinates for all of them
* there will be two different pom.xml (the one in the car file, and a second generated by Herd)

**Comment by quintesse** [comment]
> Therefore a system which depends on the module author (optionally) annotating the module at compile time is unlikely to work well in practice.

I don't know why not. I see two situations:

 - the author of the module doesn't care about Maven, doesn't (intend to) publish the module to Maven Central so anyone trying to access the module using Maven either has to go through the Herd Maven bridge or has to publish it to some private Maven repo. In both those cases the default group/artifact name we generate would be just fine
 - the author *does* care about Maven and wants to publish it on Maven Central. They do all the work and make sure they annotate their module correctly

We should definitely *not* support some weird situation where a *user* of the module decides to publish it on Maven Central!

**Comment by quintesse** [comment]
@bjansen hmmm not a bad idea at all!

Although honestly I don't like much the fact that 2 (or N, for each time the author decides to change the coordinates in the Herd) different versions of all CARs would exist. That would make checking CRCs a lot harder ("I have 2 sha1s here, one of them should probably check out")

**Comment by bjansen** [comment]
Do we really need two versions? We can rename the CAR to JAR and not modify the pom.xml inside, Maven will not use anyway (it will use the sibling pom). And I suppose the maven layout will reside in a completely different folder, right?

**Comment by bjansen** [comment]
Of course this does not solve the problem of uploading things to Maven central, because Maven coordinates won't be known locally. We *could* add a "publish to maven central" button in Herd, though.

**Comment by thradec** [comment]
I think, that to have 2 different pom for one car is pretty mess.

2016-03-09 14:40 GMT+01:00 Bastien Jansen <notifications@github.com>:

**Comment by bjansen** [comment]
Can someone please answer this?

> For what purpose is the pom.xml in the CAR generated by Ceylon in the first place?

**Comment by quintesse** [comment]
> Maven will not use anyway

If that's true then okay. It might be weird to see that Ceylon's metamodel would return different names than the one you asked for but we could probably live with that.

**Comment by quintesse** [comment]
> Can someone please answer this?

A far as I know because that's what many (all?) Maven artifacts do too, so I guess it's used *somewhere*. But I think @FroMage will know better

**Comment by gavinking** [comment]
> we're talking about serving a Maven repo from Herd, so why not configure the group id/artifact id in Herd directly? 

Right, that's exactly what I was getting at in my last comment. Stef has already decided to generate a pom using Herd itself (rather than just extracting it from the car). So it doesn't seem unreasonable to have Herd handle this stuff as well.

**Comment by FroMage** [comment]
That's possible, but it does not make it easier for people to upload to Maven Central without Herd in the equation.

**Comment by gavinking** [comment]
It's not really clear to me how and when and why the "upload to Maven Central" requirement snuck into this particular issue. Originally we were talking about letting Herd act as a mvn repo.

**Comment by ckulenkampff** [comment]
Maybe it helps to list use cases and possible problems:

1. As a Java developer, I want to develop a Java Gradle project and want to use dependencies only available on Herd.

*Currently, I can add the Herd repository as Maven repository in Gradle. This works fine for now.*

2. As a Java developer, I want to develop a Java Gradle project that uses dependencies only available on Herd and some Maven Central dependencies.

*Currently, I can add both, a Maven repository and the Herd repository, to my Gradle project. This works fine for now.*

3. As a Java developer, I want to develop a Java Gradle project that uses a Ceylon library only available at Herd with a dependency to `com.fasterxml.jackson.core`. My project also uses a big Java library which depends on the Maven variant `com.fasterxml.jackson.core:jackson-core`.

*Currently,  I can add both, a Maven repository and the Herd repository, to my Gradle project. But the dependency will be downloaded two times with different names and pollutes my classpath.*

There are similar use cases for Ceylon developers where different names should result in multiple downloads and imports of the same library.

One can always resolve these issues manually, but it makes Ceylon Herd look like an offender.

**Comment by ckulenkampff** [comment]
Just an update: Tested it with a local Sonatype Nexus installation. I could add the Herd repository as proxy. I think the following configuration was necessary to make it work better:
![image](https://cloud.githubusercontent.com/assets/923197/13730937/25d8906a-e95e-11e5-9769-232ef66a6944.png)

**Comment by luolong** [comment]
Now, there's been so much discussion here hat I might just have skimmed over some relevant bits. If so, excuse me if I rehash something that has already been decided.

About my suggestion for setting Maven `groupId:artifactId` identifiers pair --
The way I see this is that we have to have a sensible default that does not make people surprised if they do the default thing.

Then we should have some way of overriding these defaults. First on the command line and then for perpetuating the chosen setting in a configuration file somewhere (`.ceylon/config` seems good place)

#### The defaults

The most sensible default behavior I can come up with is to have group id and artifact id to be same as module id.

I know that this runs counter to how Ceylon modules have been published to maven so far, but if this is configurable later, I still think that this is the most sensible default.

#### Command line

One should be able to configure maven `groupId` and `artifactId` on the command line. In its simplest form, we could just set those up as plain text values. If a project has just one module, this should be simple enough:

```bash
ceylon compile --maven-groupId=ceylon-lang
```

The above command should produce set of modules whose `groupId` is set to `ceylon-lang` and `artifactId` is full module name

```bash
ceylon compile --maven-groupid=ceylon-lang --maven-artifactid=interop-java ceylon.interop-java
```

would produce ceylon module with maven coordinates 'ceylon-lang:interop-java:1.2.2'

Trying to call the latter command line with multiple modules (either explicitly on command line, or implicitly when calling without module name in a project with multiple modules) should be considered an error.

As an additional consideration, we could add support for some simplified regular expression syntax for deriving groupId and artifactId from the module name. Something along these lines:

```bash
ceylon compile --maven-groupid=ceylon-lang --maven-artifactId="ceylon.(**)|$1|r/./-"
```

That would take everything after "ceylon." in the module name and use it as Maven artifactId ... 
The exact syntax of the regular expression is kind of irrelevant at this point, but it should be simple enough to be able to do limited set of operations on the initial input string (module name):

* Select a portion of a string (preferably being aware of dots separating segments of module name)
* replace one string (or character) with another (e.g. replacing dots with dashes)

#### Project configuration

Whatever we allow on command line we should be able to configure in `.ceylon/config` file. With additional freedom of being able to override settings on per-module basis.

I guess the simplest configuration might be like this:

```config
[defaults]
encoding=UTF-8
overrides=build/overrides.xml

maven-groupid=ceylon-lang
maven-artifactId=ceylon.(**)|$1
...

[module "ceylon.interop.java"]
maven-artifactId=interop-java
```

**Comment by vietj** [comment]
this seems to be a great debate, however I haven't see anything about cyclic dependencies.

how are cyclic module dependencies going to be exposed by Herd to Maven ?

**Comment by luolong** [comment]
I guess we have several layers of maven/ivy interop issues under discussion here ... maybe we should split those discussions off to their own separate issues?
